File: LeetCode/Search/Rust/Leetcode719-find-k-th-smallest-pair-distance.rs
impl Solution {
    pub fn smallest_distance_pair(mut nums: Vec<i32>, k: i32) -> i32 {
        nums.sort();
        let (mut l, mut r) = (0, 1e6 as i32);
        let get = |mid: i32| -> i32 {
            let mut res = 0;
            let mut j = 0;
            for i in 0..nums.len() {
                while nums[i] - nums[j] > mid {
                    j += 1;
                }
                res += i - j;
            }
            res as i32
        };
        while l < r {
            let mid = (l + r) >> 1;
            if get(mid) >= k {
                r = mid;
            } else {
                l = mid + 1;
            }
        }
        return r;
    }
}

File: LeetCode/Search/python/Leetcode719-find-k-th-smallest-pair-distance.py
class Solution:
    def smallestDistancePair(self, nums: List[int], k: int) -> int:
        l, r = 0, int(1e6)
        nums.sort()

        def get(mid):
            res, j = 0, 0
            for i in range(len(nums)):
                while nums[i] - nums[j] > mid:
                    j += 1
                res += i - j
            return res

        while l < r:
            mid = l + r >> 1
            if get(mid) >= k: r = mid
            else: l = mid + 1
        return r
